app/services/foreman/provision/ssh.rb

```diff
@@ -50,64 +71,72 @@ def command_prefix
     (username == "root") ? "" : "sudo "
   end
 
-  def command
-    # Use the users home to store the provision script since we can't reliably
-    # tell if other locations are writeable or executable by the user.
+  # Use the users home to store the provision script since we can't reliably
+  # tell if other locations are writeable or executable by the user.
+  def main_command
     main_execution = "(chmod 0701 ./#{remote_script} && #{command_prefix} ./#{remote_script} ; echo $? >#{remote_script}.status) 2>&1"
-    "#{command_prefix} sh -c '#{main_execution} | tee #{remote_script}.log; exit $(cat #{remote_script}.status)'"
-  end
-
-  def defaults
-    {
-      :keys_only    => true,
-      :config       => false,
-      :auth_methods => %w(publickey),
-      :compression  => true,
-      :logger       => logger,
-    }
+    cleanup = "#{command_prefix} rm -f #{remote_script}"
+    "#{command_prefix} sh -c '#{main_execution} | tee #{remote_script}.log; #{cleanup}; exit $(cat #{remote_script}.status)'"
   end
 
   def logger
     Rails.logger
   end
 
-  def initiate_connection!
-    Timeout.timeout(Setting[:ssh_timeout].to_i) do
-      Timeout.timeout(8) do
-        ssh.run('pwd')
-      end
-    rescue Errno::ECONNREFUSED
-      logger.debug "Connection refused for #{address}, retrying"
-      sleep(2)
-      retry
-    rescue Errno::EHOSTUNREACH
-      logger.debug "Host unreachable for #{address}, retrying"
-      sleep(2)
-      retry
-    rescue Net::SSH::Disconnect
-      logger.debug "Host dropping connections for #{address}, retrying"
-      sleep(2)
-      retry
-    rescue Net::SSH::ConnectionTimeout
-      logger.debug "Host timed out for #{address}, retrying"
-      sleep(2)
-      retry
-    rescue Net::SSH::AuthenticationFailed
-      logger.debug "Auth failed for #{username} at #{address}, retrying"
-      sleep(2)
-      retry
-    rescue Timeout::Error
-      retry
-    rescue => e
-      Foreman::Logging.exception("SSH error", e)
-    end
+  def scp
+    logger.debug "about to upload #{template} to remote system at #{remote_script}"
+    logger.debug scp_cmd
+    _stdout, stderr, status = Open3.capture3(*scp_cmd)
+
+    return if status.success?
+
+    logger.error stderr
+    raise Foreman::Exception.new("Failed to upload '#{@template}' to remote system at #{address}")
   end
 
-  def ssh
-    Fog::SSH.new(address, username, options.merge(:timeout => 4))
+  def passwd_cmd
+    return if @options[:password].empty?
+    ["sshpass", "-p", @options[:password]]
```

Comment:
Please don't pass a password on the command line because it will be visible in the process list. Instead you can use `-e` and set `SSHPASS` as the environment variable.

This is also what REX does: https://github.com/theforeman/smart_proxy_remote_execution_ssh/blob/ced5e821124f284c9f57ee1693227c9747df715e/lib/smart_proxy_remote_execution_ssh/multiplexed_ssh_connection.rb#L36-L37

Reply:
What's the difference between `sshpass -p password` and `SSHPASS=password sshpass -e`? Both of them are exposing the value of the password

Reply:
If you run `ps` (or look in `/proc`) you can see how the process was called, including all arguments. That means any user with shell access can see it. Environment variables are private and thus secure. Generally that's why you don't want to pass passwords on the command line and SSH may run for a while so it's very easy to get the value